### PR TITLE
Add functionality to OCEX pallet to support proxy accounts and LinkedList like Storage for workers

### PR DESCRIPTION
**Pull request by Gauthamastro**
Implements a LinkedList like Storage for easy retrieval of main accounts by worker

Reference:  https://github.com/Polkadex-Substrate/polkadexTEE-worker/issues/1#issuecomment-830595079

Additional Dispatchables 
- Add proxy
- Remove proxy

### CODE REVIEW

File: pallets/ocex/src/lib.rs [hunk 1]
```diff
@@ -98,18 +129,118 @@ decl_module! {
...
+            let sender: T::AccountId = ensure_signed(origin)?;
+            ensure!(!<MainAccounts<T>>::contains_key(&sender), Error::<T>::AlreadyRegistered);
+            Self::register_acc(sender.clone())?;
+            Self::deposit_event(RawEvent::MainAccountRegistered(sender));
+            Ok(())
+        }
+
+        /// Add Proxy
+        #[weight = 10000]
+        pub fn add_proxy(origin, proxy: T::AccountId) -> DispatchResult{
```
Comment (ksr30) [inline pallets/ocex/src/lib.rs:149]: I think we also have to ensure that the given proxy is not added to some other main account. For that, we also have to maintain superset in storage i.e. ProxyAccount -> MainAccount

Comment (Gauthamastro) [reply pallets/ocex/src/lib.rs:149]: As Alain described, when an order is placed the tx contains main account that should be used. Also this open the possibility of having the same proxy for multiple main accounts. 

Example,
A market maker using two main accounts for trading but uses the same proxy ( aka trading bot).

It's a feature.

Comment (ksr30) [reply pallets/ocex/src/lib.rs:149]: hmm ... got it

**Review by ksr30 — APPROVED** [review]

File: pallets/ocex/src/lib.rs [hunk 2]
```diff
@@ -98,18 +129,118 @@ decl_module! {
...
         pub fn withdraw(origin, asset_id:  AssetId, to: T::AccountId,amount: T::Balance) -> DispatchResult{
-            let sender: T::AccountId = ensure_signed(origin)?;
+            let _: T::AccountId = ensure_signed(origin)?;
             Self::deposit_event(RawEvent::TokenWithdrawn(asset_id, to, amount));
             Ok(())
         }
 
+        /// Register MainAccount
+        #[weight = 10000]
+        pub fn register(origin) -> DispatchResult{
```
Comment (brenzi) [inline pallets/ocex/src/lib.rs:139]: what is the use case for tgis register function? Why not register proxies directly or deposit funds?

**Review by brenzi — CHANGES_REQUESTED** [review]
I'd strongly suggest to write unit tests for adding and removing proxies and main accounts in all possible ways.

The linked list approach does solve our issue of getting read proofs for all proxies nicely! Startup may take a while for a large user base, but I think this is the best option for now

File: pallets/ocex/src/lib.rs (hunk 2 above)
Comment (Gauthamastro) [reply pallets/ocex/src/lib.rs:139]: Orderbook is only allowed for those who has done decentralized KYC using KILT protocol. We will add some extra logic here that do that check

**Comment by Gauthamastro** [comment]
> I'd strongly suggest to write unit tests for adding and removing proxies and main accounts in all possible ways.
> 
> The linked list approach does solve our issue of getting read proofs for all proxies nicely! Startup may take a while for a large user base, but I think this is the best option for now
> 

Yes, since it's verified we will write the RPC and unit tests

File: pallets/ocex/src/lib.rs [hunk 3]
```diff
@@ -63,9 +84,19 @@ decl_event!(
...
+        AlreadyRegistered,
+        NotARegisteredMainAccount,
+        ProxyLimitReached
     }
 }
 
+decl_storage! {
+    trait Store for Module<T: Config> as OCEX {
+        pub FirstAccount: Option<T::AccountId> = None;
+        pub LastAccount: Option<T::AccountId> = None;
```
Comment (Gauthamastro) [inline pallets/ocex/src/lib.rs:96]: Removing first Account, use"0x00000..000"  for the first account

**Review by haerdib — APPROVED** [review]
Only a minor remark. I'll leave it up to you if you want to address it.

File: pallets/ocex/src/test.rs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            next: None,
+            proxies: vec![]
+        };
+        let linked_account: LinkedAccount<u64> = LinkedAccount{
+            prev: gen_account,
+            next: Some(new_account_two),
+            proxies: vec![]
+        };
+        let expected_linked_account_gen: LinkedAccount<u64> = LinkedAccount{
+            prev: 0,
```
[the remark this answers is not in the document]
Comment (ksr30) [reply pallets/ocex/src/test.rs:51]: Hey @haerdib, you are right ... I forgot to change it

**Pull request merged**